File: src/secretsharer.cpp
//
//  secretsharer.cpp
//  SmcsClient.ArrayBpXorEncoding
//
//  Created by James Sinclair on 06/02/2014.
//  Copyright (c) 2014 James Sinclair. All rights reserved.
//

#include <stdio.h>
#include <stdlib.h>

#ifdef WINDOWS
    #include <direct.h>
    #define GetCurrentDir _getcwd
#else
    #include <unistd.h>
    #define GetCurrentDir getcwd
 #endif

#include <string>
#include <libgen.h>
#include <iostream>
#include <fstream>
#include "boost/filesystem.hpp"
#include "blocksecretsharer.hh"

#include <unistd.h>
#include <sys/types.h>
#include <pwd.h>


using namespace std;

// Prints help-message
void print_help(bool unknown_command);

// Create a new key
void create_key();

// Create shares of a file
bool create_file_shares(string filepath, string output_message);

// Reconstruct a file from shares
bool reconstruct_file_from_shares(string filepath, string output_message);

// Gets contents of a file, based upon file-path
unsigned char * get_file_contents(string filepath, int * size_read);

// Puts a buffer to a file, based upon
void put_file_contents(string filepath, unsigned char * contents, int length);

// Gets current path of the running exe
int get_current_path(char * current_path, int length);

// Creates a share-name based upon the input filename
string create_share_name(string filename, int share_index);

struct passwd *pw = getpwuid(getuid());
const string homedir = pw->pw_dir;
const string _key_file_name = "/.ssk";
const string _key_file_path = homedir + _key_file_name;

const int _max_message_length = 100;


/* The format for this command-line tool is as follows -
 *  
 *  secretsharer -h                 : displays the help-mesage
 *  secretsharer -k                 : create new key. Keys will be
 *                                  : - stored in file 
 *                                  :   ./.key
 *
 *  secretsharer -c <file-path>     : creates shares from the file
 *                                  : - identified by <file-path>
 *
 *  secretsharer -r <file-path>     : reconstructs a file from shares
 *                                  : - of a file based on the format
 *                                  : - of <file-path>. Format will 
 *                                  : - be simple -
 *                                  :       <file-path>_%
 *                                  : - where % is the index of the 
 *                                  : - share-file
 */
int main(int argc, char * argv[])
{
    string output_message;
    int readbytes = 0;
    unsigned char * bytesread;
    string arg2;
    
    // Check there are enough arguments
    if (argc == 2 && argv[1][0] == '-')     // argument must begin with a dash, '-'
    {
        switch (argv[1][1])
        {
            case 'h':
                print_help(false);
                break;
                
            case 'k':
                create_key();
                break;
                
            default:
                print_help(true);
                break;
        }
    }
    // first argument only must begin with a dash, '-'
    else if (argc == 3 && argv[1][0] == '-' && argv[2][0] != '-')
    {
	arg2 = argv[2];
        switch (argv[1][1])
        {
            case 'c':
                create_file_shares(arg2, output_message);
                break;
                
            case 'r':
                reconstruct_file_from_shares(arg2, output_message);
                break;
                
            case 'g':
                bytesread = get_file_contents(arg2, &readbytes);
                cout << "\nget_file_contents operation read " << readbytes << " bytes from file '" << arg2 << "'\n";
                cout << "bytes read: '" << bytesread << "'\n";
                break;
                
            default:
                print_help(true);
                break;
        }
        
    }
    // first argument only must begin with a dash, '-'
    else if (argc == 4 && argv[1][0] == '-' && argv[2][0] != '-' && argv[3][0] != '-')
    {
	arg2 = argv[2];
        switch (argv[1][1])
        {
            case 'p':
                put_file_contents(arg2, (unsigned char *)argv[3], sizeof(argv[3]));
                break;
                
            default:
                print_help(true);
                break;
        }
    }
    else
    {
        // Print help message
        print_help(argc != 1);
    }
    
    return (0);
}


// Prints help-message
void print_help(bool unknown_command)
{
    if (unknown_command)
    {
        cout << "UNKNOWN COMMAND\n\n";
    }

    cout << "\n\nThe format for this command-line tool is as follows -\n";

    cout << "  secretsharer -h                 : displays the help-mesage\n";
    cout << "  secretsharer -k                 : create new key. Keys will be\n";
    cout << "                                  : - stored at\n";
    cout << "                                  :   ./.key\n\n";
    cout << "  secretsharer -c <file-path>     : creates shares from the file\n";
    cout << "                                  : - identified by <file-path>\n\n";
    cout << "  secretsharer -r <file-path>     : reconstructs a file from shares\n";
    cout << "                                  : - of a file based on the format\n";
    cout << "                                  : - of <file-path>. Format will\n";
    cout << "                                  : - be simple -\n";
    cout << "                                  :       <file-path>_%\n";
    cout << "                                  : - where % is the index of the\n";
    cout << "                                  : - share-file\n";

}

// Create a new key
void create_key()
{
    //cout << "\ncreate_key() called\n\n";
    
    unsigned char * buffer;
    int length = BlockSecretSharer::KeyLength();

    ISecretSharer * secretSharer = new BlockSecretSharer();

    buffer = secretSharer->Key();

    put_file_contents(_key_file_path, buffer, length);

    delete secretSharer;
}

// Create shares of a file
bool create_file_shares(string file_path, string output_message)
{
    bool success = false;
    
    unsigned char * buffer;
    int length = 0;
    string share_name;

    // Ensure input-file exists
    buffer = get_file_contents(file_path, &length);


    if (length > 0)
    {
	// Get key and set up sharer
	BlockSecretSharer * sharer = NULL;
	int read_size = 0;
	unsigned char * key = get_file_contents(_key_file_path, &read_size);
	if (read_size > 0)
	{
	    sharer = new BlockSecretSharer(key);
	    delete[] key;
	}
	else
	{
	    sharer = new BlockSecretSharer();
	}
	
	map<int, vector<unsigned char> > shares = sharer->CreateShares(buffer, length, 4);
	
	delete[] buffer;
	delete sharer;

	for (int i = 1; i <= 4; i++)
	{
	    share_name = create_share_name(file_path, i);
	    put_file_contents(share_name, shares[i].data(), length);
	}
	success = true;
    }
    
    return success;
}

// Reconstruct a file from shares
bool reconstruct_file_from_shares(string file_path, string output_message)
{
    bool success = false;

    // Look for locally-stored key
    BlockSecretSharer * sharer = NULL;
    int read_size = 0;
    unsigned char * key = get_file_contents(_key_file_path, &read_size);
    if (read_size > 0)
    {
	sharer = new BlockSecretSharer(key);	// Key found, use for this operation
	delete[] key;

	map<int, vector<unsigned char> > shares;
	unsigned char * buffer;
	int length;

	// Get at least 2 shared files from local directory
	for (int i = 1; i <= 4; i++)
	{
	    length = 0;
	    string share_name = create_share_name(file_path, i);
	    buffer = get_file_contents(share_name, &length);

	    if (length > 0)
	    {
		shares[i] = vector<unsigned char>(buffer, buffer + length);
		delete[] buffer;
		if (shares.size() >= 2)
		    break;
	    }
	}
	
	if (shares.size() >= 2)
	{
	    // Found at least 2 shared files from local directory. Do reconstruction
	    unsigned char * reconstructed_file = sharer->ReconstructSecret(shares, length, length);

	    // Write reconstructed data to original filename.
	    put_file_contents(file_path, reconstructed_file, length);
	    
	    delete[] reconstructed_file;
	    
	    success = true;
	}
	else
	{
	    cout << "\n\nNo shared files found.\n\n";
	}
	delete sharer;
    }
    else
    {
	cout << "\n\nNo key found, reconstruction would fail!\n\n";
	return false;
    }
    
    return success;
}

// Gets contents of a file, based upon file-path, returns length of buffer
unsigned char * get_file_contents(string file_path, int * read_size)
{
    
    streampos size;
    char * memblock;
    
    ifstream file (file_path.c_str(), ios::in|ios::ate|ios::binary); // Set position to end of file
    if (file.is_open())
    {
        size = file.tellg();        // Position is at the end of file, so gives us file-size
        memblock = new char[size];
        file.seekg (0, ios::beg);
        file.read(memblock, size);
        file.close();
        
        *read_size = size;
    }
    else cout << "\n\nUnable to open file '" << file_path << "'\n\n";
    
    return (unsigned char *)memblock;
}

// Puts a buffer to a file, based upon file-path
void put_file_contents(string file_path, unsigned char * contents, int length)
{
    //cout << "\nput_file_contents() called\n\n";
    
    char * memblock;
    
    ofstream file (file_path.c_str(), ios::out|ios::binary);
    if (file.is_open())
    {
        memblock = (char *)contents;
        file.write(memblock, length);
        file.close();
    }
    else cout << "\n\nUnable to open file '" << file_path << "'\n\n";
}

// Gets current path of the running exe
int get_current_path(string current_path)
{
    int length = current_path.length();
    char * c_str = new char [length+1];
    strcpy(c_str, current_path.c_str());

    if (!GetCurrentDir(c_str, length))
    {
	return errno;
    }
    return (0);
}

// Creates a share-name based upon the input filename
string create_share_name(string filename, int share_index)
{
    string sharename;

    // Split extension and basename from filename
    std::string::size_type pos;
    string extensionmarker(".");
    pos = filename.find_last_of(extensionmarker);
    char char_share_index = '0' + share_index;

    if (pos == string::npos || pos == 0)
    {
	// No match found or dot found at first position, use filename as basename
	sharename = filename;
    }
    else
    {
	// Get basename from filename to position
	sharename = filename.substr(0, pos);
    }

    // Add share-index to basename
    sharename += "_";
    sharename += char_share_index;

    // Add extension to new basename and copy to sharename
    sharename += filename.substr(pos);

    return sharename;
}

File: PythonImplementation/secretsharer.py
#!/usr/bin/python2
"""A command-line tool for sharing file-data into obscured shares"""

from blocksecretsharer import BlockSecretSharer
import sys
import os.path
import io
#import pdb


KEY_FILE_NAME = "~/.ssk"


def print_help(unknown_command, command=None):
    """Prints the help-screen, listing all valid arguments."""
    if unknown_command is True:
        print "\nUNKNOWN COMMAND:", command

    print "\nThe format for this command-line tool is as follows -\n"

    print "  secretsharer -h                 : displays the help-message"
    print "  secretsharer -k                 : create new key. Keys will be"
    print "                                  : - stored at"
    print "                                  :    ", KEY_FILE_NAME
    print "  secretsharer -c <file-path>     : creates shares from the file"
    print "                                  : - identified by <file-path>"
    print "  secretsharer -r <file-path>     : reconstructs a file from shares"
    print "                                  : - of a file based on the format"
    print "                                  : - of <file-path>. Format will"
    print "                                  : - be simple -"
    print "                                  : -     <file-path>_%"
    print "                                  : - where % is the index of the"
    print "                                  : - share-file"


def create_key():
    """Creates a new encryption-key and stores in in home-directory."""

    #Get new key
    sharer = BlockSecretSharer()
    key = sharer.get_key()

    #Write key to file
    put_file_contents(os.path.expanduser(KEY_FILE_NAME), key)

    del sharer


def get_key():
    """Gets currently held encryption-key stored in home directory."""

    return get_file_contents(os.path.expanduser(KEY_FILE_NAME))


def create_file_shares(filepath):
    """Creates shares of a file at the given path."""

    success = False

    contents = get_file_contents(filepath)
    length = len(contents)

    if length > 0:
        #Get key and set up sharer
        key = get_key()

        if len(key) > 0:
            #Create secret-sharer with current key
            sharer = BlockSecretSharer(key)
        else:
            #Create secret-sharer with new key
            sharer = BlockSecretSharer()

        #Create shares from file-contents
        shares = sharer.create_shares(contents)

        #Write shared data to files
        number_of_shares = sharer.get_number_of_shares()
        for i in range(1, number_of_shares + 1):
            #Get share-name
            share_name = create_share_name(filepath, i)
            #Write share-data to file
            put_file_contents(share_name, shares[i])

        del sharer

        success = True

    return success


def reconstruct_file_from_shares(filepath):
    """Reconstructs a file at a given path from shares.
    The share-paths are inferred from the file-path."""

    success = False

    #Look for locally-stored key
    key = get_key()

    if len(key) > 0:
        #Key found, use for this operation
        sharer = BlockSecretSharer(key)

        shares = {}
        number_of_shares = sharer.get_number_of_shares()

        #Get at least 2 shared files from same directory as filepath
        for i in range(1, number_of_shares + 1):

            share_name = create_share_name(filepath, i)
            contents = get_file_contents(share_name)

            if len(contents) > 0:
                shares[i] = contents
                if len(shares) >= 2:
                    break

        if len(shares) >= 2:
            #Found at least 2 shared files from directory local to filepath.
            #Do reconstruction
            reconstructed_file = sharer.reconstruct_secret(shares)

            #Write reconstructed data to original filename.
            put_file_contents(filepath, reconstructed_file)

            success = True
        else:
            print "\nNo shared files found"

        del sharer

    else:
        print "\nNo key found, reconstruction would fail!"

    return success


def get_file_contents(filepath):
    """Gets the contents of a file at a given path."""

    file_contents = None

    if os.path.isfile(filepath):
        try:
            input_file = io.open(filepath, 'rb')
            file_contents = bytearray(input_file.read())
        except IOError:
            print "\nError: can\'t find file or read data from file at", \
                filepath
        except io.BlockingIOError:
            print "\nError: unsuported operation while reading data from file at", \
                filepath
        except:
            print "\nError: unknown error occurred while opening file for", \
                "reading at", filepath
        else:
            input_file.close()

    return file_contents


def put_file_contents(filepath, contents):
    """Writes given data (as a list) to a file at the given path."""

    try:
        #Convert data from list to array of bytes
        contents_bytes = bytearray(contents)
        output_file = io.open(filepath, 'wb')
        output_file.write(contents_bytes)
        output_file.flush()
    except IOError:
        print "\nError: can\'t find file or write data to file at", \
            filepath
    except io.BlockingIOError:
        print "\nError: unsuported operation while writing data to file at", \
            filepath
    except:
        print "\nError: unknown error occurred while opening file for", \
            "writing at", filepath
    else:
        output_file.close()


def create_share_name(filename, share_index):
    """Creates a standard share-name from a given filename and share-index."""

    sharename = ""

    if "." in filename:
        filename_parts = filename.split('.')
        num_parts = len(filename_parts)
        for i in range(0, num_parts - 1):
            sharename += filename_parts[i]
        sharename += "_"
        sharename += str(share_index)
        sharename += "."
        sharename += filename_parts[num_parts - 1]
    else:
        sharename = filename + "_" + share_index

    return sharename


def main_function(argv):
    """Main"""

    num_of_args = len(argv)

    #Check there are enough arguments
    if num_of_args == 2 and argv[1][0] == '-':

        if argv[1] == "-h":
            print_help(False)

        elif argv[1] == "-k":
            create_key()

        else:
            print_help(True, argv[1])

    elif num_of_args == 3 and argv[1][0] == '-' and argv[2][0] != '-':

        if argv[1] == "-c":
            create_file_shares(argv[2])

        elif argv[1] == "-r":
            reconstruct_file_from_shares(argv[2])

        elif argv[1] == "-g":
            bytes_read = get_file_contents(argv[2])
            print "get_file_contents operation read", str(len(bytes_read)), \
                "bytes from file", argv[2]

        else:
            print_help(True, argv[1])

    elif num_of_args == 4 and argv[1][0] == '-' and argv[2][0] != '-':

        if argv[1] == "-p":
            put_file_contents(argv[2], argv[3])

    else:

        print_help(False, argv[1:])


if __name__ == "__main__":

    main_function(sys.argv)
